### PR TITLE
stylized layer icons in legend (future data-menu)

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request generalizes the existing stylized layer icons used for layer tabs in plot options to also be used in the viewer layer legends (which will eventually be used to open the new data-menu).

Screenshot 2024-10-16 at 8 34 49 AM

Screenshot 2024-10-16 at 8 34 59 AM

TODO:
- [x] fix test failures
- [x] re-implement prefix_icon n the flyout

Known "delayed update" issues hopefully fixed by upstream subset messaging work:
- linewidth in spectrum viewer legend (updates after change to color)
- new spectral subset in spectrum viewer legend (updates after additional subset)
- change in color_mode needs to force styling update (changing color/colormap after does)

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by kecnry** [comment]
This is now ready for reviews (see notes in description about known buggy behaviors which likely will be follow-up efforts unless anyone has concerns).  I'm leaving this in draft since we want to defer merge until after the 4.0 release.

File: jdaviz/configs/default/plugins/data_menu/data_menu.vue [hunk 1]
```diff
@@ -1,43 +1,74 @@
...
   </div>
 </template>
 
 <style scoped>
   .viewer-label {
     display: block;
     float: right;
-    background-color: #c3c3c3c3;
-    width: 24px;
+    background-color: #c3c3c32c;
```
Comment (bmorris3) [inline jdaviz/configs/default/plugins/data_menu/data_menu.vue:57]: Is this change intentional?

File: jdaviz/components/layer_viewer_icon_stylized.vue
Comment (bmorris3) [inline jdaviz/components/layer_viewer_icon_stylized.vue:1]: Should we specify `cursor: default` until the icons are clickable (and then `pointer`)?

File: jdaviz/components/layer_viewer_icon_stylized.vue [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    colorBackgroundStyle(colors, cmap_samples) {
+      const strip_width = 42 / colors.length
+      var cmap_strip_width = strip_width
+      var style_colors = []
+      var style = 'repeating-linear-gradient( 135deg, '
+  
+      for ([mi, color_or_cmap] of colors.entries()) {
+        if (color_or_cmap == 'from_list') {
+          /* follow-up: use actual colors from the DQ plugins */
+          color_or_cmap = 'rainbow'
```
Comment (bmorris3) [inline jdaviz/components/layer_viewer_icon_stylized.vue:69]: I'm not sure what color DQ should get. We could take the first N colors from the DQ colormap like you're probably saying in this comment, but they're (intentionally) random, and this might not be the most visually meaningful way to represent DQ.

File: jdaviz/components/layer_viewer_icon_stylized.vue [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      if (visible === 'mixed'){
+        return 'repeating-linear-gradient(30deg, rgba(0,0,0,0.3), rgba(0,0,0,0.3) 3px, transparent 3px, transparent 3px, transparent 10px)'
+      }
+      else if (visible) {
+        return 'repeating-linear-gradient(30deg, transparent, transparent 10px)'
+      } else {
+        return 'repeating-linear-gradient(30deg, rgba(0,0,0,0.4), rgba(0,0,0,0.4) 8px, transparent 8px, transparent 8px, transparent 10px)'
+      }
+    },
+    colorBackgroundStyle(colors, cmap_samples) {
```
Comment (bmorris3) [inline jdaviz/components/layer_viewer_icon_stylized.vue:60]: I don't expect that these loops to evaluate the colormap for the icons cost much. But is there a way to prove to ourselves that this is true?

**Review by bmorris3 — COMMENTED** [review]
Looks good to me so far (given the caveats listed in the description).

File: jdaviz/configs/default/plugins/data_menu/data_menu.vue (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/data_menu/data_menu.vue:57]: yes, this adds transparency to the existing gray

File: jdaviz/components/layer_viewer_icon_stylized.vue
Comment (kecnry) [reply jdaviz/components/layer_viewer_icon_stylized.vue:1]: good idea, done via passing `disabled` to the underlying `v-btn`.

File: jdaviz/components/layer_viewer_icon_stylized.vue (hunk 3 above)
Comment (kecnry) [reply jdaviz/components/layer_viewer_icon_stylized.vue:61]: we could try to profile it?  All this is doing though is taking the input lists generated in python and joining them into a string, with a bit of math to set the widths of each entry, so I really don't expect it to show up in anything.

**Comment by rosteen** [comment]
I'm pretty sure I know what mixed visibility means in this case (the subset layer for only one data layer is visible), but I could see it being confusing for the user considering they can see the subset no matter what data layer is shown. Not specific to this PR as it's also this way in Plot Options, but it's more...visible...here, and thus might cause more confusion. Would it perhaps make more sense to not have the cross hatching on subset layers? Again, not a hold up for this PR, just the first thing I thought when testing this.

Screenshot 2024-10-18 at 5 48 09 PM

**Comment by rosteen** [comment]
Doing more testing this morning, I found a bug in Specviz. The layer icon for a subset only appears after you make another subset, rather than when the first is finalized:


https://github.com/user-attachments/assets/7734972c-e0f4-4f6f-b31b-4bc27890ac95

**Comment by rosteen** [comment]
Never mind, that was a known issue.

**Review by rosteen — APPROVED** [review]
Looks ok to me, modulo the known issues to be handled in follow ups.

**Review by bmorris3 — APPROVED** [review]
This looks good to go. Thanks!
